fxrecorder/src/lib/analysis.rs

```diff
@@ -0,0 +1,406 @@
+    Ok(frames_dir)
+}
+
+/// Information about a frame being processed in
+/// [`find_first_orange_frame`][function.find_first_orange_frame.html].
+#[derive(Debug)]
+struct FrameInfo {
+    /// The path to the frame.
+    path: PathBuf,
+
+    /// The frame number.
+    frame_num: u32,
+}
+
+/// Squared Euclidean Distance between two colours as 3-vectors.
+fn squared_distance(a: &Rgb<u8>, b: &Rgb<u8>) -> i64 {
+    let dr = a[0] as i64 - b[0] as i64;
+    let dg = a[1] as i64 - b[1] as i64;
+    let db = a[2] as i64 - b[2] as i64;
+
+    dr * dr + dg * dg + db * db
+}
+
+#[derive(Debug, Error)]
+pub enum OrangeError {
+    #[error("could not read frame directory: {}", .0)]
+    ReadDir(#[source] io::Error),
+
+    #[error("could not read file `{}': {}", .1.display(), .0)]
+    Open(#[source] io::Error, PathBuf),
+
+    #[error("could not load image `{}': {}'", .1.display(), .0)]
+    Load(#[source] ImageError, PathBuf),
+
+    #[error("no orange frame detected")]
+    MissingOrange,
+}
+
+/// Return the frame number of the first orange frame of the video.
+fn find_first_orange_frame(log: slog::Logger, frames_dir: &Path) -> Result<u32, OrangeError> {
+    // This is the orange that Splash generates and that visuametrics.py expects.
+    let orange = image::Rgb([222, 100, 13]);
+
+    let mut frames = vec![];
+    for entry in read_dir(frames_dir).map_err(OrangeError::ReadDir)? {
+        let entry = entry.map_err(OrangeError::ReadDir)?;
+        let path = entry.path();
+        let path_str = String::from(path.file_name().unwrap().to_str().unwrap());
+
+        let suffix_pos = match path_str.find(".png") {
+            Some(pos) => pos,
+            None => {
+                warn!(
+                    log,
+                    "unexpected non-PNG file found in frames directory; skipping";
+                    "filename" => path.display()
+                );
+                continue;
+            }
+        };
+
+        let frame_num: u32 = match path_str[..suffix_pos].parse() {
+            Ok(n) => n,
+            Err(_) => {
+                warn!(
+                    log,
+                    "unexpected PNG file with non-numeric filename; skipping";
+                    "filename" => path.display(),
+                );
+                continue;
+            }
+        };
+
+        frames.push(FrameInfo { path, frame_num });
+    }
+
+    frames.sort_by(|a, b| a.frame_num.cmp(&b.frame_num));
+
+    for info in &frames {
+        let f = BufReader::new(
+            File::open(&info.path)
+                .map_err(|source| OrangeError::Open(source, info.path.clone()))?,
+        );
+        let image = image::load(f, image::ImageFormat::Png)
+            .map_err(|source| OrangeError::Load(source, info.path.clone()))?
+            .into_rgb();
+
+        let avg =
+            average_image(&image.view(image.width() / 2 - 25, image.height() / 2 - 25, 50, 50));
+        if squared_distance(&avg, &orange) < 500 {
```

Comment:
Nit: can you pull the magic numbers here out into constants? They can be locally defined, just if they are named it would help a little bit with readability.